sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +129,24 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    case in @ In(Cast(fromExp, toType: NumericType, _), list)
+      if list.forall(v =>
+        v.isInstanceOf[Literal] && canImplicitlyCast(fromExp, toType, v.dataType)
```

Comment:
The analyzer makes sure that the list of `In` is already of the same data type. Here we can simply check
```
list.forall(_.isInstanceOf[Literal]) && canImplicitlyCast(fromExp, toType, list.head.dataType)
```

Reply:
updated.

Reply:
Sorry, made some mistake. We should check the `v.dataType` in `in.list`, due to we may get a `fromExpr` that is not in `NumericType`